gsoc2020/20200303-health-check-wjayesh.md

```diff
@@ -0,0 +1,254 @@
+
+Thus, making the state of CoreDNS available externally and reliably is important to ensure important services run as they are expected to. A shell script could’ve been used to automate the steps but shell is not good in error handling and may not be reliable depending on which shell (bash/csh/zsh) to use. A Golang program will be much more robust.  
+ 
+### Deliverables  
+
+
+- A standalone Golang application to check CoreDNS health: 
+  - The binary delivered through this project will be a **flexible** application in that it could choose its approach to the problem depending on the environment it is deployed in and the access that has been provided to it.   
+  - The health check program itself doesn’t necessarily reside in the Kubernetes cluster. 
+  - This binary only concerns itself with the health of CoreDNS deployed on Kubernetes. It doesn’t cover other scenarios. 
+
+
+- Proper documentation for the product explaining steps and the concepts used. 
+ 
+ 
+ 
+
+ 
+## Design Proposal 
+
+
+It is necessary to consider all possible cases because the way any team deploys the binary is their **choice** and we should not be forcing a modus operandi to use the product.  
+ 
+Minimal assumptions about the users’ Kubernetes deployment are made and the primary **focus** is to build a program that could easily be adopted by more teams (which may have different preferences of deployment). 
+ 
+In any case, the binary will accomplish the following fundamental tasks: 
+
+
+- Use `kubectl` to extract information about the cluster and its services and pods.  
+ 
+  This will help determine the procedure to follow and also facilitate the subsequent actions to take. This step can be completed only when the application can connect and communicate with the API server. The Go [client](https://github.com/kubernetes/client-go) for Kubernetes has packages that can discover and authenticate with the API server.  
+ 
+- Find the IP address (cluster IP and pod IP) of the CoreDNS service and pods.  
+ 
+  These IP addresses will be used for the `dig` operation in the next steps. It is critical to use both cluster IP and the pod IP. Although it may seem that pinging cluster IP would be sufficient, in real production systems, any point of the system could fail. We cannot make the assumption that Kubernetes networking (forwarding requests from the Service IP to the Pods) will work all the time.  
+ 
+  Apart from this, it’ll be useful to know which Pods are sick, so they can be remedied individually.  
+ 
+- Use `dig` to ping the cluster IP and pod IPs through DNS. 
+ 
+  `dig` is a tool that can be used to look up DNS servers using their IP addresses and port numbers (default is 53). The output of the command execution will help determine whether the DNS service is healthy or not.  
+ 
+- Restart CoreDNS pods through the API server if pods are up and running but DNS reply does not match the expectation. 
+
+
+ 
+### Strategies for Different Deployment Scenarios 
```

Comment:
@WJayesh, please change to "Deployment Strategies", "Deployment Scenarios", etc. Also, do you really need a subsection here? I don't see any other subsections. Alternatively, you may need to add a few, e.g. "Design", "Timeline", etc.

Reply:
Should I include the content of "Deployment Scenarios" inside Implementation? I'll also add the Kubernetes API references, use of `kubectl` and the golang library references when describing the points.